### PR TITLE
Add Half for sparse.mm reduce

### PR DESCRIPTION
**Pull request by yanbing-j**
This PR is to add Half support for sparse.mm reduce in CPU backend.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: test/test_sparse_csr.py [hunk 1]
```diff
@@ -2649,7 +2649,7 @@ def ref(row, col, val, mat):
             out = torch.sparse.mm(csr, mat, reduce_type)
             self.assertEqual(out, ref_out)
 
-            if train and dtype is not torch.bfloat16:
+            if train and dtype not in [torch.bfloat16, torch.float16]:
```
[the remark this answers is not in the document]
Comment (yanbing-j) [reply test/test_sparse_csr.py]: Done. Thanks!

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -13576,7 +13576,7 @@ def sample_inputs_alias_copy(op_info, device, dtype, requires_grad, **kwargs):
                DecorateInfo(unittest.skip("Skipped!"), 'TestFakeTensor', 'test_fake_crossref_backward_no_amp'),
            )),
     OpInfo('sparse.mm',
-           dtypes=floating_types_and(torch.bfloat16),
+           dtypes=floating_types_and(torch.bfloat16, torch.float16),
```
Comment (Skylion007) [inline torch/testing/_internal/common_methods_invocations.py:13579]: This is more of a general question, but why do we only run sparse_mm tests on CPU, don't we also support sparse.mm on CUDA to some extend too now? Should we remove the `onlyCPU` decorators and args to OpInfo?

File: test/inductor/test_torchinductor_opinfo.py [hunk 3]
```diff
@@ -226,7 +226,7 @@ def format_op(op):
     ("normal", "in_place"): {f16, f32, f64},
     ("normal", "number_mean"): {f16, f32, f64},
     "normal": {f16, f32, f64},
-    ("sparse.mm", "reduce"): {f32, f64},
+    ("sparse.mm", "reduce"): {f32, f64, f16},
```
Comment (Skylion007) [inline test/inductor/test_torchinductor_opinfo.py:229]:
```suggestion
    ("sparse.mm", "reduce"): {f32, f64, f16, bf16},
```

**Review by Skylion007 — APPROVED** [review]
For extending sparse CPU support, this is fine, but we may need to improve our OpInfo tests in this area.

**Comment by Skylion007** [comment]
I tried the demo in the doc:
```
import torch
a = torch.tensor([[1., 0, 2], [0, 3, 0]], device="cuda").to_sparse().requires_grad_()
a
b = torch.tensor([[0, 1.], [2, 0], [0, 0]], device="cuda", requires_grad=True)
b
y = torch.sparse.mm(a, b)
y
y.sum().backward()
a.grad
c = a.detach().to_sparse_csr()
c
y1 = torch.sparse.mm(c, b, 'sum')
y1
y2 = torch.sparse.mm(c, b, 'max')
y2
``` 
crashed as soon as 'sum' and 'max' was specified and sparse support is still in beta set, so probably why we don't have good optests for it yet.

**Pull request closed without merging**